text/0000-const-locals.md

```diff
@@ -0,0 +1,68 @@
+- Feature Name: const_locals
+- Start Date: 2018-01-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow `let` bindings in the body of constants and const fns. Additionally enable
+destructuring in `let` bindings and const fn arguments.
+
+# Motivation
+[motivation]: #motivation
+
+It makes writing const fns much more like writing regular functions and is
+not possible right now because the old constant evaluator was a constant folder
+that could only process expressions. With the miri const evaluator this feature
+exists but is still disallowed.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`let` bindings in constants and const fn work just like `let` bindings
+everywhere else. Historically these did not exist in constants and const fn
+because it would have been very hard to support them in the old const evaluator.
+
+This means that you can only move out of any let binding once, even though in a
+const environment obtaining a copy of the object could be done by executing the
+code twice, side effect free. All invariants held by runtime code are also
```

Comment:
`const fn` are also callable from outside of a constant context, so it's not always true that their arguments can be reproduced. +1 for the RFC's approach.